kolibri/core/content/management/commands/importcontent.py

```diff
@@ -285,11 +286,10 @@ def _transfer(  # noqa: max-complexity=16
         with self.start_progress(
             total=total_bytes_to_transfer + dummy_bytes_for_annotation
         ) as overall_progress_update:
-            exception = None  # Exception that is not caught by the retry logic
-
             if method == DOWNLOAD_METHOD:
                 session = requests.Session()
 
+            file_transfers = []
```

Comment:
Hrm... there's something about precalculating all the file transfer objects that is bothering me. Feels like it might add quite a bit to the memory usage for downloads of large numbers of file - although I don't know how much of an impact it would be.

One possibility would be to turn this transfer generation into a generator function on the command class, and then instead of using `executor.submit` we can use `executor.map` to apply the `_start_file_transfer` function over the generator iterable. This should mean that we only need to have as many file transfer objects at once as we are downloading, which then makes it no worse than the previous implementation in this regard. Thoughts?

Reply:
Did some more searching online about how people deal with this, and the solution that seems to work the best is dividing the list of actions into batches. I've updated the code to do that, and now it never goes above 250MB total RAM usage. Surprisingly, I'm seeing an average of 45m instead of 50m for importing KA now too, so it may also improve performance a bit?

Reply:
Win win!